File: src/feature/frame/frameModel.ts
import { v4 as uuidv4 } from "uuid";

export function generateId(): string {
  return uuidv4();
}

export class BlockModel {
  id: string;
  frameId: string;
  keyType: string;
  key: string;
  visibilityKey: string;
  position: number;
  slot: string | null;
  integrationVersion: number;
  parentId: string | null;
  blocks?: BlockModel[] | null;
  data: BlockDataModel[] | null;
  properties: BlockPropertyModel[];
  slots: BlockSlotModel[];
  actions: ActionModel[];

  constructor(
    frameId: string,
    keyType: string,
    key: string,
    visibilityKey: string,
    position: number,
    slot: string | null,
    integrationVersion: number,
    parentId: string | null,
    blocks: BlockModel[] | null,
    data: BlockDataModel[],
    properties: BlockPropertyModel[],
    slots: BlockSlotModel[],
    actions: ActionModel[]
  ) {
    this.id = generateId();
    this.frameId = frameId;
    this.keyType = keyType;
    this.key = key;
    this.visibilityKey = visibilityKey;
    this.position = position;
    this.slot = slot;
    this.integrationVersion = integrationVersion;
    this.parentId = parentId;
    this.blocks = blocks;
    this.data = data;
    this.properties = properties;
    this.slots = slots;
    this.actions = actions;
  }
}

export class BlockPropertyModel {
  id: string;
  key: string;
  value?: string;
  valueMobile?: string;
  valueTablet?: string;
  valueDesktop?: string;
  type: string;
  description?: string;
  valuePicker?: string;
  valuePickerGroup?: string;
  valuePickerOptions?: string;
  blockId: string;

  constructor(
    blockId: string,
    key: string,
    type: string,
    value?: string,
    valueMobile?: string,
    valueTablet?: string,
    valueDesktop?: string,
    description?: string,
    valuePicker?: string,
    valuePickerGroup?: string,
    valuePickerOptions?: string
  ) {
    this.id = generateId();
    this.key = key;
    this.value = value;
    this.valueMobile = valueMobile;
    this.valueTablet = valueTablet;
    this.valueDesktop = valueDesktop;
    this.type = type;
    this.description = description;
    this.valuePicker = valuePicker;
    this.valuePickerGroup = valuePickerGroup;
    this.valuePickerOptions = valuePickerOptions ?? "[]";
    this.blockId = blockId;
  }
}

export class BlockDataModel {
  id: string;
  blockId: string;
  key: string;
  value: string;
  type: string;
  description?: string;

  constructor(blockId: string, key: string, value: string, type: string, description?: string) {
    this.id = generateId();
    this.blockId = blockId;
    this.key = key;
    this.value = value;
    this.type = type;
    this.description = description;
  }
}

export class BlockEventModel {
  eventName: string;
  actions: any[];

  constructor(eventName: string, actions: any[] = []) {
    this.eventName = eventName;
    this.actions = actions;
  }
}

export class BlockSlotModel {
  id: string;
  blockId: string;
  slot: string;
  description?: string;

  constructor(blockId: string, slot: string, description?: string) {
    this.id = generateId();
    this.blockId = blockId;
    this.slot = slot;
    this.description = description;
  }
}

export interface ActionModel {
  id: string;
  frameId: string;
  key: string;
  event: string;
  triggers: ActionTriggerModel[];
}

export class ActionTriggerModel {
  id: string;
  actionId: string;
  parentId: string | null;
  keyType: string;
  then: string;
  name: string;
  integrationVersion: number;
  properties: TriggerPropertyModel[];
  data: TriggerDataModel[];
  triggers: ActionTriggerModel[];

  constructor(
    actionId: string,
    parentId: string | null,
    keyType: string,
    then: string,
    name: string,
    integrationVersion: number,
    properties: TriggerPropertyModel[],
    data: TriggerDataModel[],
    triggers: ActionTriggerModel[]
  ) {
    this.id = generateId();
    this.actionId = actionId;
    this.parentId = parentId;
    this.keyType = keyType;
    this.then = then;
    this.name = name;
    this.integrationVersion = integrationVersion;
    this.properties = properties;
    this.data = data;
    this.triggers = triggers;
  }
}

export class TriggerPropertyModel {
  id: string;
  actionTriggerId: string;
  key: string;
  value?: string;
  type: string;
  description?: string;
  valuePicker?: string;
  valuePickerGroup?: string;
  valuePickerOptions?: string;

  constructor(
    actionTriggerId: string,
    key: string,
    type: string,
    value?: string,
    description?: string,
    valuePicker?: string,
    valuePickerGroup?: string,
    valuePickerOptions?: string
  ) {
    this.id = generateId();
    this.actionTriggerId = actionTriggerId;
    this.key = key;
    this.value = value;
    this.type = type;
    this.description = description;
    this.valuePicker = valuePicker;
    this.valuePickerGroup = valuePickerGroup;
    this.valuePickerOptions = valuePickerOptions ?? "[]";
  }
}

export class TriggerDataModel {
  id: string;
  actionTriggerId: string;
  key: string;
  value: string;
  type: string;
  description?: string;

  constructor(actionTriggerId: string, key: string, value: string, type: string, description?: string) {
    this.id = generateId();
    this.actionTriggerId = actionTriggerId;
    this.key = key;
    this.value = value;
    this.type = type;
    this.description = description;
  }
}

File: cmd/frame/frameModel.go
package frame

type FrameModel struct {
	ID             string          `json:"id"`
	Name           string          `json:"name"`
	Route          string          `json:"route"`
	RouteArguments []RouteArgument `json:"routeArguments"`
	Type           string          `json:"type"`
	IsStarter      bool            `json:"isStarter"`
	ProjectID      string          `json:"projectId"`
	Checksum       string          `json:"checksum"`
	Variables      []VariableModel `json:"variables"`
	Blocks         []BlockModel    `json:"blocks"`
	Actions        []ActionModel   `json:"actions"`
}

type RouteArgument struct {
	Name string `json:"name"`
}

type VariableModel struct {
	ID      string `json:"id"`
	FrameID string `json:"frameId"`
	Key     string `json:"key"`
	Value   string `json:"value"`
	Type    string `json:"type"`
}

type BlockModel struct {
	ID                 string               `json:"id"`
	FrameID            string               `json:"frameId"`
	KeyType            string               `json:"keyType"`
	Key                string               `json:"key"`
	VisibilityKey      string               `json:"visibilityKey"`
	Position           int                  `json:"position"`
	Slot               *string              `json:"slot,omitempty"`
	IntegrationVersion int                  `json:"integrationVersion"`
	ParentID           *string              `json:"parentId,omitempty"`
	Data               []BlockDataModel     `json:"data"`
	Properties         []BlockPropertyModel `json:"properties"`
	Slots              []BlockSlotModel     `json:"slots"`
}

type BlockPropertyModel struct {
	ID                 string  `json:"id"`
	BlockID            string  `json:"blockId"`
	Key                string  `json:"key"`
	ValueMobile        string  `json:"valueMobile"`
	ValueTablet        string  `json:"valueTablet"`
	ValueDesktop       string  `json:"valueDesktop"`
	Type               string  `json:"type"`
	Description        *string `json:"description,omitempty"`
	ValuePicker        string  `json:"valuePicker"`
	ValuePickerGroup   string  `json:"valuePickerGroup"`
	ValuePickerOptions string  `json:"valuePickerOptions"`
}

type BlockDataModel struct {
	ID          string  `json:"id"`
	BlockID     string  `json:"blockId"`
	Key         string  `json:"key"`
	Value       string  `json:"value"`
	Type        string  `json:"type"`
	Description *string `json:"description,omitempty"`
}

type BlockSlotModel struct {
	ID          string  `json:"id"`
	BlockID     string  `json:"blockId"`
	Slot        string  `json:"slot"`
	Description *string `json:"description,omitempty"`
}

type ActionModel struct {
	ID       string               `json:"id"`
	FrameID  string               `json:"frameId"`
	Key      string               `json:"key"`
	Event    string               `json:"event"`
	Triggers []ActionTriggerModel `json:"triggers"`
}

type ActionTriggerModel struct {
	ID                 string                 `json:"id"`
	ActionID           string                 `json:"actionId"`
	ParentID           *string                `json:"parentId,omitempty"`
	KeyType            string                 `json:"keyType"`
	Then               string                 `json:"then"`
	Name               string                 `json:"name"`
	IntegrationVersion int                    `json:"integrationVersion"`
	Properties         []TriggerPropertyModel `json:"properties"`
	Data               []TriggerDataModel     `json:"data"`
}

type TriggerPropertyModel struct {
	ID                 string  `json:"id"`
	ActionTriggerID    string  `json:"actionTriggerId"`
	Key                string  `json:"key"`
	Value              string  `json:"value"`
	Type               string  `json:"type"`
	Description        *string `json:"description,omitempty"`
	ValuePicker        *string `json:"valuePicker,omitempty"`
	ValuePickerGroup   *string `json:"valuePickerGroup,omitempty"`
	ValuePickerOptions *string `json:"valuePickerOptions,omitempty"`
}

type TriggerDataModel struct {
	ID              string  `json:"id"`
	ActionTriggerID string  `json:"actionTriggerId"`
	Key             string  `json:"key"`
	Value           string  `json:"value"`
	Type            string  `json:"type"`
	Description     *string `json:"description,omitempty"`
}
